README.md
# Nested redirects for Laravel

[![Build Status](https://travis-ci.org/Neurony/laravel-redirects.svg?branch=master)](https://travis-ci.org/Neurony/laravel-redirects)
[![StyleCI](https://github.styleci.io/repos/189837919/shield?branch=master)](https://github.styleci.io/repos/189837919)
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/Neurony/laravel-redirects/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/Neurony/laravel-redirects/?branch=master)

- [Overview](#overview)   
- [Installation](#installation)   
- [Usage](#usage)   

# Overview

This package allows you to create simple or multiple nested redirects for your Laravel applications.   
   
This package can be useful from an SEO perspective, when in your application, you have URLs that have the potential of being modified.
   
**Example of the dynamic redirecting logic:**
* Let's assume you have an URL called `/original`   
   
* You create a redirect from `/original` to `/modified`
  > Accessing `/original` will redirect to `/modified`   
* You create another redirect from `/modified` to `/modified-again`   
  > Accessing `/modified` will redirect to `/modified-again` AND   
  > Accessoing `/original` will redirect to `/modified-again`   
* You create another redirect from `/modified-again` to `/modified-yet-again`   
  > Accessing `/modified-again` will redirect to `/modified-yet-again` AND      
  > Accessing `/modified` will redirect to `/modified-yet-again` AND   
  > Accessing `/original` will redirect to `/modified-yet-again`   
* You create another redirect from `modified-yet-again` to `/original`  
  > Accessing `/modified-yet-again` will redirect to `/original` AND   
  > Accessing `/modified-again` will redirect to `/original` AND   
  > Accessing `/modified` will redirect to `/original`
  
# Installation

Install the package via Composer:

```
composer require neurony/laravel-redirects
```

Publish the config file with:

```
php artisan vendor:publish --provider="Neurony\Redirects\ServiceProvider" --tag="config"
```

Publish the migration file with:

```
php artisan vendor:publish --provider="Neurony\Redirects\ServiceProvider" --tag="migrations"
```

After the migration has been published you can create the `redirects` table by running:

```
php artisan migrate
```

# Usage

### Add the middleware

In order for the redirecting functionality to actually happen, you need to add the `Neurony\Redirects\Middleware\RedirectRequests` middleware.

Go to `App\Http\Kernel` and add the `Neurony\Redirects\Middleware\RedirectRequests` middleware in your `$middlewareGroups` groups of choice.

```php
/**
 * The application's route middleware groups.
 *
 * @var array
 */
protected $middlewareGroups = [
    'web' => [
        ...
        \Neurony\Redirects\Middleware\RedirectRequests::class,
```

### Creating redirects

You should never use the `Neurony\Redirects\Models\Redirect` directly, as this is the default concrete implementation for the `Neurony\Redirects\Contracts\RedirectModelContract`.   
  
Using the `Neurony\Redirects\Models\Redirect` model class directly will prevent you from being able to extend the model's capabilities.

You can create redirects that will be stored inside the `redirects` table like this:   

```php
app('redirect.model')->create([
    'old_url' => '/your-old-url',
    'new_url' => '/your-new-url',
    'status' => 301
]);
```

To see how you can extend the `Neurony\Redirects\Models\Redirect` model's capabilities, please read the comments from `/config/redirects.php -> redirect_model`

# Credits

- [Andrei Badea](https://github.com/zbiller)
- [All Contributors](../../contributors)

# Security

If you discover any security related issues, please email andrei.badea@neurony.ro instead of using the issue tracker.

# License

The MIT License (MIT). Please see [LICENSE](LICENSE.md) for more information.

# Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

# Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.


CHANGELOG.md
# Changelog

All notable changes to `laravel-redirects` will be documented in this file  

## 2.0.0 - 2020/10/06

- Support Laravel 7.x 

## 2.0.0 - 2019/13/11

- Support Laravel 6.0

## 1.0.0 - 2019/6/2

- Initial release


tests/TestCase.php
<?php

namespace Neurony\Redirects\Tests;

use Illuminate\Contracts\Foundation\Application;
use Illuminate\Contracts\Http\Kernel;
use Neurony\Redirects\Middleware\RedirectRequests;
use Orchestra\Testbench\TestCase as Orchestra;

abstract class TestCase extends Orchestra
{
    /**
     * Setup the test environment.
     *
     * @return void
     * @throws \Illuminate\Contracts\Container\BindingResolutionException
     */
    public function setUp(): void
    {
        parent::setUp();

        $this->setUpDatabase($this->app);
        $this->setUpMiddleware($this->app);
    }

    /**
     * Register the service provider.
     *
     * @param \Illuminate\Foundation\Application $app
     * @return array
     */
    protected function getPackageProviders($app)
    {
        return [ServiceProvider::class];
    }

    /**
     * Define environment setup.
     *
     * @param Application $app
     */
    protected function getEnvironmentSetUp($app)
    {
        $app['config']->set('database.default', 'testing');
        $app['config']->set('database.connections.sqlite', [
            'driver' => 'sqlite',
            'database' => ':memory:',
            'prefix' => '',
        ]);
    }

    /**
     * Set up the database and migrate the necessary tables.
     *
     * @param  $app
     */
    protected function setUpDatabase(Application $app)
    {
        $this->loadMigrationsFrom(__DIR__.'/Database/Migrations');
    }

    /**
     * Register the middleware.
     *
     * @param Application $app
     * @throws \Illuminate\Contracts\Container\BindingResolutionException
     */
    public function setUpMiddleware(Application $app)
    {
        $app->make(Kernel::class)->pushMiddleware(RedirectRequests::class);
    }
}


tests/RedirectTest.php
<?php

namespace Neurony\Redirects\Tests;

use Neurony\Redirects\Exceptions\RedirectException;
use Neurony\Redirects\Models\Redirect;

class RedirectTest extends TestCase
{
    /** @test */
    public function it_redirects_a_request()
    {
        Redirect::create([
            'old_url' => 'old-url',
            'new_url' => 'new/url',
        ]);

        $response = $this->get('old-url');
        $response->assertRedirect('new/url');
    }

    /** @test */
    public function it_redirects_nested_requests()
    {
        Redirect::create([
            'old_url' => '1',
            'new_url' => '2',
        ]);

        $response = $this->get('1');
        $response->assertRedirect('2');

        Redirect::create([
            'old_url' => '2',
            'new_url' => '3',
        ]);

        $response = $this->get('1');
        $response->assertRedirect('3');

        $response = $this->get('2');
        $response->assertRedirect('3');

        Redirect::create([
            'old_url' => '3',
            'new_url' => '4',
        ]);

        $response = $this->get('1');
        $response->assertRedirect('4');

        $response = $this->get('2');
        $response->assertRedirect('4');

        $response = $this->get('3');
        $response->assertRedirect('4');

        Redirect::create([
            'old_url' => '4',
            'new_url' => '1',
        ]);

        $response = $this->get('2');
        $response->assertRedirect('1');

        $response = $this->get('3');
        $response->assertRedirect('1');

        $response = $this->get('4');
        $response->assertRedirect('1');
    }

    /** @test */
    public function it_guards_against_creating_redirect_loops()
    {
        $this->expectException(RedirectException::class);

        Redirect::create([
            'old_url' => 'same-url',
            'new_url' => 'same-url',
        ]);
    }
}


tests/ServiceProvider.php
<?php

namespace Neurony\Redirects\Tests;

use Illuminate\Contracts\Foundation\Application;
use Illuminate\Support\ServiceProvider as BaseServiceProvider;
use Neurony\Redirects\Contracts\RedirectModelContract;
use Neurony\Redirects\Models\Redirect;

class ServiceProvider extends BaseServiceProvider
{
    /**
     * Create a new service provider instance.
     *
     * @param Application $app
     */
    public function __construct(Application $app)
    {
        parent::__construct($app);
    }

    /**
     * Bootstrap the application services.
     *
     * @return void
     * @throws \Illuminate\Contracts\Container\BindingResolutionException
     */
    public function boot()
    {
        $this->publishConfigs();
    }

    /**
     * Register the application services.
     *
     * @return void
     */
    public function register()
    {
        $this->registerBindings();
    }

    /**
     * Setup the config.
     *
     * @throws \Illuminate\Contracts\Container\BindingResolutionException
     */
    protected function publishConfigs()
    {
        $this->app->make('config')->set('redirects.statuses', [
            302 => 'Normal (302)',
            301 => 'Permanent (301)',
            307 => 'Temporary (307)',
        ]);
    }

    /**
     * @return void
     */
    protected function registerBindings()
    {
        $this->app->bind(RedirectModelContract::class, $this->config['redirects']['redirect_model'] ?? Redirect::class);
        $this->app->alias(RedirectModelContract::class, 'redirect.model');
    }
}


src/Models/Redirect.php
<?php

namespace Neurony\Redirects\Models;

use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Model;
use Neurony\Redirects\Contracts\RedirectModelContract;
use Neurony\Redirects\Exceptions\RedirectException;

class Redirect extends Model implements RedirectModelContract
{
    /**
     * The database table.
     *
     * @var string
     */
    protected $table = 'redirects';

    /**
     * The attributes that are mass assignable.
     *
     * @var array
     */
    protected $fillable = [
        'old_url',
        'new_url',
        'status',
    ];

    /**
     * Boot the model.
     *
     * @return void
     */
    public static function boot()
    {
        parent::boot();

        static::saving(function (self $model) {
            if (trim(strtolower($model->old_url), '/') == trim(strtolower($model->new_url), '/')) {
                throw RedirectException::sameUrls();
            }

            static::whereOldUrl($model->new_url)->whereNewUrl($model->old_url)->delete();

            $model->syncOldRedirects($model, $model->new_url);
        });
    }

    /**
     * The mutator to set the "old_url" attribute.
     *
     * @param string $value
     */
    public function setOldUrlAttribute($value)
    {
        $this->attributes['old_url'] = trim(parse_url($value)['path'], '/');
    }

    /**
     * The mutator to set the "new_url" attribute.
     *
     * @param string $value
     */
    public function setNewUrlAttribute($value)
    {
        $this->attributes['new_url'] = trim(parse_url($value)['path'], '/');
    }

    /**
     * Filter the query by an old url.
     *
     * @param Builder $query
     * @param string $url
     *
     * @return mixed
     */
    public function scopeWhereOldUrl($query, string $url)
    {
        return $query->where('old_url', $url);
    }

    /**
     * Filter the query by a new url.
     *
     * @param Builder $query
     * @param string $url
     *
     * @return mixed
     */
    public function scopeWhereNewUrl($query, string $url)
    {
        return $query->where('new_url', $url);
    }

    /**
     * Get all redirect statuses defined inside the "config/redirects.php" file.
     *
     * @return array
     */
    public static function getStatuses(): array
    {
        return (array) config('redirects.statuses', []);
    }

    /**
     * Sync old redirects to point to the new (final) url.
     *
     * @param RedirectModelContract $model
     * @param string $finalUrl
     * @return void
     */
    public function syncOldRedirects(RedirectModelContract $model, string $finalUrl): void
    {
        $items = static::whereNewUrl($model->old_url)->get();

        foreach ($items as $item) {
            $item->update(['new_url' => $finalUrl]);
            $item->syncOldRedirects($model, $finalUrl);
        }
    }

    /**
     * Return a valid redirect entity for a given path (old url).
     * A redirect is valid if:
     * - it has an url to redirect to (new url)
     * - it's status code is one of the statuses defined on this model.
     *
     * @param string $path
     * @return Redirect|null
     */
    public static function findValidOrNull($path): ?RedirectModelContract
    {
        return static::where('old_url', $path === '/' ? $path : trim($path, '/'))
            ->whereNotNull('new_url')
            ->whereIn('status', array_keys(self::getStatuses()))
            ->latest()->first();
    }
}


src/ServiceProvider.php
<?php

namespace Neurony\Redirects;

use Illuminate\Contracts\Foundation\Application;
use Illuminate\Routing\Router;
use Illuminate\Support\Facades\File;
use Illuminate\Support\Facades\Route;
use Illuminate\Support\ServiceProvider as BaseServiceProvider;
use Neurony\Redirects\Contracts\RedirectModelContract;
use Neurony\Redirects\Middleware\RedirectRequests;
use Neurony\Redirects\Models\Redirect;

class ServiceProvider extends BaseServiceProvider
{
    /**
     * @var Router
     */
    protected $router;

    /**
     * Create a new service provider instance.
     *
     * @param Application $app
     */
    public function __construct(Application $app)
    {
        parent::__construct($app);
    }

    /**
     * Bootstrap the application services.
     *
     * @param Router $router
     * @return void
     */
    public function boot(Router $router)
    {
        $this->router = $router;

        $this->publishConfigs();
        $this->publishMigrations();
        $this->registerMiddleware();
        $this->registerRouteBindings();
    }

    /**
     * Register the application services.
     *
     * @return void
     */
    public function register()
    {
        $this->registerBindings();
    }

    /**
     * @return void
     */
    protected function publishConfigs()
    {
        $this->publishes([
            __DIR__.'/../config/redirects.php' => config_path('redirects.php'),
        ], 'config');
    }

    /**
     * @return void
     */
    protected function publishMigrations()
    {
        if (empty(File::glob(database_path('migrations/*_create_redirects_table.php')))) {
            $timestamp = date('Y_m_d_His', time());
            $migration = database_path("migrations/{$timestamp}_create_redirects_table.php");

            $this->publishes([
                __DIR__.'/../database/migrations/create_redirects_table.php.stub' => $migration,
            ], 'migrations');
        }
    }

    /**
     * @return void
     */
    protected function registerMiddleware()
    {
        $this->router->aliasMiddleware('redirect.requests', RedirectRequests::class);
    }

    /**
     * @return void
     */
    protected function registerRouteBindings()
    {
        Route::model('redirect', RedirectModelContract::class);
    }

    /**
     * @return void
     */
    protected function registerBindings()
    {
        $this->app->bind(RedirectModelContract::class, $this->config['redirects']['redirect_model'] ?? Redirect::class);
        $this->app->alias(RedirectModelContract::class, 'redirect.model');
    }
}


src/Contracts/RedirectModelContract.php
<?php

namespace Neurony\Redirects\Contracts;

interface RedirectModelContract
{
    /**
     * @param string $value
     */
    public function setOldUrlAttribute($value);

    /**
     * @param string $value
     */
    public function setNewUrlAttribute($value);

    /**
     * @param \Illuminate\Database\Eloquent\Builder $query
     * @param string $url
     * @return mixed
     */
    public function scopeWhereOldUrl($query, string $url);

    /**
     * @param \Illuminate\Database\Eloquent\Builder $query
     * @param string $url
     * @return mixed
     */
    public function scopeWhereNewUrl($query, string $url);

    /**
     * @return array
     */
    public static function getStatuses(): array;

    /**
     * @param RedirectModelContract $model
     * @param string $finalUrl
     * @return void
     */
    public function syncOldRedirects(self $model, string $finalUrl): void;

    /**
     * @param string $path
     * @return RedirectModelContract|null
     */
    public static function findValidOrNull($path): ?self;
}


config/redirects.php
<?php

return [

    /*
    |
    | The redirect statuses that you will use in your application.
    | By default, the "301", "302" and "307" are defined.
    |
    */
    'statuses' => [
        302 => 'Normal (302)',
        301 => 'Permanent (301)',
        307 => 'Temporary (307)',
    ],

    /*
    |
    | Concrete implementation for the "redirect model".
    | To extend or replace this functionality, change the value below with your full "redirect model" FQN.
    |
    | Your class will have to (first option is recommended):
    | - extend the "Neurony\Redirects\Models\Redirect" class
    | - or at least implement the "Neurony\Redirects\Contracts\RedirectModelContract" interface.
    |
    | Regardless of the concrete implementation below, you can still use it like:
    | - app('redirect.model') OR app('\Neurony\Redirects\Contracts\RedirectsModelContract')
    | - or you could even use your own class as a direct implementation
    |
    */
    'redirect_model' => \Neurony\Redirects\Models\Redirect::class,

];
